runtime/near-vm-runner/src/prepare/prepare_v2.rs

```diff
@@ -1,58 +1,326 @@
+                            .checked_sub(u64::from(count))
+                            .ok_or(PrepareError::TooManyLocals)?;
+                    }
+
+                    let func_validator = self
+                        .validator
+                        .code_section_entry(&func)
+                        .map_err(|_| PrepareError::Deserialization)?;
+                    // PANIC-SAFETY: no big deal if we panic here while the allocations are taken.
+                    // Worst-case we are going to be making new allocations again, but in practice
+                    // this should never happen as this context should not be reused.
+                    let allocs = std::mem::replace(
+                        &mut self.func_validator_allocations,
+                        wp::FuncValidatorAllocations::default(),
+                    );
+                    let mut func_validator = func_validator.into_validator(allocs);
+                    func_validator.validate(&func).map_err(|_| PrepareError::Deserialization)?;
+                    self.func_validator_allocations = func_validator.into_allocations();
+                }
+                wp::Payload::CustomSection(reader) => {
+                    self.ensure_import_section();
+                    self.copy_section(SectionId::Custom, reader.range())?;
+                }
+
+                // Extensions not supported.
+                wp::Payload::UnknownSection { .. }
+                | wp::Payload::TagSection(_)
+                | wp::Payload::ModuleSection { .. }
+                | wp::Payload::InstanceSection(_)
+                | wp::Payload::CoreTypeSection(_)
+                | wp::Payload::ComponentSection { .. }
+                | wp::Payload::ComponentInstanceSection(_)
+                | wp::Payload::ComponentAliasSection(_)
+                | wp::Payload::ComponentTypeSection(_)
+                | wp::Payload::ComponentCanonicalSection(_)
+                | wp::Payload::ComponentStartSection { .. }
+                | wp::Payload::ComponentImportSection(_)
+                | wp::Payload::ComponentExportSection(_) => {
+                    tracing::trace!("input module contains unsupported section");
+                    return Err(PrepareError::Deserialization);
+                }
+            }
+        }
+        Ok(std::mem::replace(&mut self.output_code, Vec::new()))
+    }
+
+    fn transform_import_section(
+        &mut self,
+        reader: &wp::ImportSectionReader,
+    ) -> Result<(), PrepareError> {
+        let mut new_section = wasm_encoder::ImportSection::new();
+        for import in reader.clone() {
+            let import = import.map_err(|_| PrepareError::Deserialization)?;
+            if import.module != "env" {
+                return Err(PrepareError::Instantiate);
+            }
+            let new_type = match import.ty {
+                wp::TypeRef::Func(id) => {
+                    // TODO: validate imported function types here.
```

Comment:
We should be validating full signatures of every imported function. This is pre-existing comment from v1.

Reply:
Hm, then are we actively planning to add these checks any time soon?
Or asked another way, what are the consequences of not doing it? I assume we will catch it at a later stage anyway?

Reply:
It is a non-trivial amount of work impact of which would be comparatively less than many of the other stuff we can do, but it would definitely mean a cleaner and more proper architecture. Right now all of the imports are "validated" when linking the module, which means that such failing deploys are much more expensive than they could be and I guess it also runs a risk of forward compatibility hazards (i.e. adding any new host functions right now requires a protocol version bump just in case anybody has deployed a then-invalid contract that would become valid with the new host function in place.)

Reply:
That makes a lot of sense, thanks for the explanation!

Reply:
For the last one, I think adding new host functions must be a protocol version bump anyway whatever happens, as otherwise someone could upload a contract with this function during the version upgrade period and cause a challenge to be formed